Sources/SourceKitLSP/Swift/SyntaxHighlightingTokenParser.swift

```diff
@@ -0,0 +1,247 @@
+
+import SourceKitD
+import LanguageServerProtocol
+import LSPLogging
+
+/// Parses tokens from sourcekitd response dictionaries.
+struct SyntaxHighlightingTokenParser {
+  private let sourcekitd: SourceKitD
+  /// Use `keys.nameoffset` and `keys.namelength` instead of `keys.offset`
+  /// and `keys.length`. This is useful in AST structures provided by
+  /// sourcekitd, since these contain both a name range and a normal range,
+  /// where the latter refers to the entire node (e.g. the entire function
+  /// declaration).
+  private let useName: Bool
+
+  init(sourcekitd: SourceKitD, useName: Bool = false) {
+    self.sourcekitd = sourcekitd
+    self.useName = useName
+  }
+
+  func parseTokens(_ response: SKDResponseDictionary, in snapshot: DocumentSnapshot) -> [SyntaxHighlightingToken] {
+    let keys = sourcekitd.keys
+    var tokens: [SyntaxHighlightingToken] = []
+
+    if let offset: Int = useName ? response[keys.nameoffset] : response[keys.offset],
+       var length: Int = useName ? response[keys.namelength] : response[keys.length],
+       let start: Position = snapshot.positionOf(utf8Offset: offset),
+       let skKind: sourcekitd_uid_t = response[keys.kind],
+       case (let kind, var modifiers)? = parseKindAndModifiers(skKind) {
+
+      // We treat function declaration and enum member name tokens as a special
+      // case, e.g. SourceKit returns `f(x: Int, y: Int)` as a name instead of just `f`.
+      if useName && [.function, .method, .enumMember].contains(kind) && modifiers.contains(.declaration),
+         let name: String = response[keys.name],
+         name.contains("("),
+         let funcNameLength: Int = name.split(separator: "(").first?.utf8.count {
+        length = funcNameLength
+      }
+
+      // If the name is escaped in backticks, we need to add two characters to the
+      // length for the backticks.
+      if modifiers.contains(.declaration),
+         let index = snapshot.indexOf(utf8Offset: offset), snapshot.text[index] == "`" {
+        length += 2
+      }
+
+      if let isSystem: Bool = response[keys.is_system], isSystem {
+        modifiers.insert(.defaultLibrary)
+      }
+
+      if let end: Position = snapshot.positionOf(utf8Offset: offset + length) {
+        let multiLineRange = start..<end
+        let ranges = multiLineRange.splitToSingleLineRanges(in: snapshot)
+
+        tokens += ranges.map {
+          SyntaxHighlightingToken(
+            range: $0,
+            kind: kind,
+            modifiers: modifiers
+          )
+        }
+      }
+    }
+
+    if let substructure: SKDResponseArray = response[keys.substructure] {
+      tokens += parseTokens(substructure, in: snapshot)
+    }
+
+    return tokens
+  }
+
+  func parseTokens(_ response: SKDResponseArray, in snapshot: DocumentSnapshot) -> [SyntaxHighlightingToken] {
+    var result: [SyntaxHighlightingToken] = []
+    response.forEach { (_, value) in
+      result += parseTokens(value, in: snapshot)
+      return true
+    }
```

Comment:
Should we pass in `result` as an inout parameter to avoid the intermediate array?